### PR TITLE
JEI Sortable Panel Integration and other improvements.

### PR DESCRIPTION
**Pull request by IMarvinTPA**
Add some internal sorting for Weapons, Armor, and Tools.
Add the ability to filter on ClassName.
Add the ability to filter on ToolClass and ArmorType.
Add the ability for the tree to have neighboring nodes to share the same sort index to allow other sorting to take place.
Integrate with JEI's customizable sortable item panel.
Add alternate API to unsort items that are not matched by the tree.
Updated Forge version.
Updated Item Tree XML for modded items.

For Issues #404, #555, #556

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -1,9 +1,9 @@
 mod_group=inventorytweaks
 mod_name=InventoryTweaks
-mod_version=1.64-dev
+mod_version=1.65-dev
```
Comment (Kobata) [inline gradle.properties]: I don't see a final 1.64 anywhere, so this is fairly unnecessary.

File: src/main/java/invtweaks/InvTweaks.java [hunk 2]
```diff
@@ -448,16 +468,180 @@ int compareItems(@NotNull ItemStack i, @NotNull ItemStack j, int orderI, int ord
...
+        		int maxDamage1 = i.getMaxDamage() <= 0 ? Integer.MAX_VALUE : i.getMaxDamage();
+        		int maxDamage2 = j.getMaxDamage() <= 0 ? Integer.MAX_VALUE : j.getMaxDamage();
+        		if (maxDamage2 != maxDamage1)
+        			return maxDamage2 - maxDamage1;  
+
+        		//Use remaining durability to sort, favoring more damaged.
+        		int curDamage1 = i.getItemDamage();
+        		int curDamage2 = j.getItemDamage();
+        		if (curDamage2 != curDamage1) {
+                    if(i.isItemStackDamageable() && !getConfigManager().getConfig().getProperty(InvTweaksConfig.PROP_INVERT_TOOL_DAMAGE).equals(InvTweaksConfig.VALUE_TRUE)) {
```
Comment (Kobata) [inline src/main/java/invtweaks/InvTweaks.java]: There seem to be some general formatting inconsistencies around all this, especially in the preview this line is indented *less* than the one before it.

File: src/main/java/invtweaks/InvTweaksItemTree.java [hunk 3]
```diff
@@ -4,13 +4,20 @@
...
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.JsonToNBT;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTUtil;
+import net.minecraft.util.NonNullList;
+import net.minecraft.util.ResourceLocation;
 import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import net.minecraftforge.fml.common.registry.ForgeRegistries;
 import net.minecraftforge.oredict.OreDictionary;
+import scala.Int;
```
Comment (Kobata) [inline src/main/java/invtweaks/InvTweaksItemTree.java]: I think this is the wrong import for that. There isn't much reason for anything from `scala` to be imported.

(Added later: Also looks unused)

File: src/main/java/invtweaks/InvTweaksItemTree.java [hunk 4]
```diff
@@ -48,8 +55,13 @@
     private String rootCategory;
     @NotNull
     private List<OreDictInfo> oresRegistered = new ArrayList<>();
+    
+    @NotNull
+    private List<ItemStack> allGameItems = new ArrayList<ItemStack>();
```
Comment (Kobata) [inline src/main/java/invtweaks/InvTweaksItemTree.java:61]: Nothing outside of here seems to ever use this, is there a reason for it to exist? It's potentially a good chunk of memory in highly-modded installs just sitting around persistently.

File: src/main/java/invtweaks/InvTweaksItemTreeLoader.java [hunk 5]
```diff
@@ -5,6 +5,8 @@
 import net.minecraft.nbt.NBTException;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraftforge.common.MinecraftForge;
+import scala.util.parsing.json.JSON;
```
Comment (Kobata) [inline src/main/java/invtweaks/InvTweaksItemTreeLoader.java]: Scala stuff again. Also appears entirely unused?

File: src/main/resources/assets/inventorytweaks/itemtree.xml [hunk 6]
```diff
@@ -16,1181 +16,10191 @@
...
-            <rabbitsFoot id="rabbit_foot"/>
-            <rabbitFoot id="rabbit_foot"/>
-        </potionIngredient>
-    </item>
-</stuff>
+	<equipment>
+		<weapon>
+			<meleeWeapon>
+				<sword>
+					<itemCreativeSword id="randomthings:creativesword"/>
```
Comment (Kobata) [inline src/main/resources/assets/inventorytweaks/itemtree.xml]: The default item tree should probably not explicitly name any modded items. I do some general oredict categories and such to catch common ones but there's way too many popular mods (and too many items in them) to make a heavy attempt to keep up on them.

Perhaps a secondary optional file containing the expanded entries would be more suitable?

**Review by Kobata — CHANGES_REQUESTED** [review]

**Comment by Kobata** [comment]
The travis issue is just it apparently being mad that PR checks don't get the signing key, I don't think there's been any (PRs) since that was added?

I pointed it out a little above but there's definitely something weird up with the formatting of all the changes: Most obviously around spaces vs tabs since that makes things align wrong on the web preview.

File: src/main/java/invtweaks/InvTweaksItemTree.java (hunk 3 above)
Comment (IMarvinTPA) [reply src/main/java/invtweaks/InvTweaksItemTree.java]: I had been using Int.MaxValue().  I've been swapping that out and forgot to remove the import.

File: src/main/java/invtweaks/InvTweaks.java (hunk 2 above)
Comment (IMarvinTPA) [reply src/main/java/invtweaks/InvTweaks.java]: I've been working on JEI and it is all-tabs-all-the-time.  It seems InvTweaks is all-space-all-the-time.  So, I'll need to keep that in mind and work out a fix.

File: src/main/java/invtweaks/InvTweaksItemTree.java (hunk 4 above)
Comment (IMarvinTPA) [reply src/main/java/invtweaks/InvTweaksItemTree.java:61]: It now clears the list when the tree is done loading and (re)creates the list when the tree is reloaded.

**Review by IMarvinTPA — COMMENTED** [review]
General comment ended up in the wrong spot.

**Comment by IMarvinTPA** [comment]
For the ItemTree, I've had hacked-up JEIs and InventoryTweaks that supported me in building the trees by seeing the whole tree at once in the inventory panel.  The current tree gets about 52% of all the items in StoneBlock.  I'm hoping by getting these integrations together, it'll free me up to do more tree maintenance rather than re-hacking.  I would hope that having a beefier tree, more people would be more comfortable with the inventory tweaks sorting and maybe even assist in improving it.

The formatting should be improved greatly.  I tried to get git filter processing working, but ultimately did a search-and-replace on tab to 4 spaces.

Thank You.

**Comment by Kobata** [comment]
I've just been generally wary of adding too much setup specifically for certain mods.

Although, like I said I think there is a place for a more complete tree but that it should be separate from the vanilla-only one: both because there are a lot of people who don't really use many other mods with this, and because it makes for a much easier to update thing for new Minecraft versions, when a lot of mods end up massively changing their items and blocks anyway.

(The ideal would probably be a massive rework of the whole tree loading system so there can be different files merged but doing that obviously wouldn't be amazingly practical at the moment)

**Comment by IMarvinTPA** [comment]
Can I add a button next to the "Edit Tree" that says "Load Default Modded Tree" that swaps them out?

**Comment by Kobata** [comment]
Yeah I guess that'd work.

File: src/main/resources/assets/inventorytweaks/lang/en_US.lang [hunk 7]
```diff
@@ -39,6 +39,8 @@ invtweaks.settings.rulesfile=Open the sorting rules file...
 invtweaks.settings.rulesfile.error=[8] Failed to open rules file
 invtweaks.settings.treefile=Open the item tree file...
 invtweaks.settings.treefile.error=[9] Failed to open tree file
+invtweaks.settings.moddedtreefile=Load modded tree file
+invtweaks.settings.moddedtreefile.error=Failed to replace tree file
```
Comment (Kobata) [inline src/main/resources/assets/inventorytweaks/lang/en_US.lang]: This should probably be prefixed with a number: [14] looks like the first unused one (and wouldn't require all the translations to be updated) even though that breaks the pattern of going from top to bottom in the file

File: src/main/java/invtweaks/InvTweaksItemTreeLoader.java [hunk 8]
```diff
@@ -18,6 +19,7 @@
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Locale;
```
Comment (Kobata) [inline src/main/java/invtweaks/InvTweaksItemTreeLoader.java]: I don't see any uses of this import in the diff.

**Review by Kobata — APPROVED** [review]

File: src/main/resources/assets/inventorytweaks/lang/en_US.lang (hunk 7 above)
Comment (IMarvinTPA) [reply src/main/resources/assets/inventorytweaks/lang/en_US.lang]: The actual number is 17 and the pattern is already broken.

File: src/main/java/invtweaks/InvTweaksItemTreeLoader.java (hunk 8 above)
Comment (IMarvinTPA) [reply src/main/java/invtweaks/InvTweaksItemTreeLoader.java]: I removed ti from the toLowerCase when I was having troubles matching stringvariable == "constant".  Removed unused import.

**Comment by Kobata** [comment]
Ok, is everything done enough to merge? I don't see anything remaining incredibly obvious.

**Comment by IMarvinTPA** [comment]
I think I'm done with the features.  I'll send a new pull request when I have done a fair bit more to the moddeditemtree file.

**Comment by Kobata** [comment]
Alright, shoved it in.

**Pull request merged**